nni/algorithms/compression/v2/pytorch/utils/evaluator.py

```diff
@@ -0,0 +1,713 @@
+            def hook(grad):
+                buffer.append(grad.clone())
+            return hook
+    """
+
+    def __init__(self, target: Tensor, target_name: str, hook_factory: Callable[[List], Callable[[Tensor], Any]]):
+        assert isinstance(target, Tensor)
+        super().__init__(target, target_name, hook_factory)
+
+    def _register(self, hook_func: Callable[[Tensor], Any]) -> RemovableHandle:
+        return self.target.register_hook(hook_func)  # type: ignore
+
+
+class ModuleHook(Hook):
+    def __init__(self, target: Module, target_name: str, hook_factory: Callable[[List], Callable[[Module, Tensor, Tensor], Any]]):
+        assert isinstance(target, Module)
+        super().__init__(target, target_name, hook_factory)
+
+
+class ForwardHook(ModuleHook):
+    """
+    Here is an example for hook_factory, in this example, the output of this module will be saved in the buffer::
+
+        def hook_factory(buffer):
+            def hook(module, input, output):
+                buffer.append(output.clone())
+            return hook
+    """
+
+    def _register(self, hook_func: Callable[[Module, Tensor, Tensor], Any]):
+        return self.target.register_forward_hook(hook_func)  # type: ignore
+
+
+class BackwardHook(ModuleHook):
+    """
+    Here is an example for hook_factory, in this example, the gradient of this module input will be saved in the buffer::
+
+        def hook_factory(buffer):
+            def hook(module, grad_input, grad_output):
+                buffer.append(grad_input.clone())
+            return hook
+    """
+
+    def _register(self, hook_func: Callable[[Module, Tensor, Tensor], Any]):
+        return self.target.register_backward_hook(hook_func)  # type: ignore
+
+
+class Evaluator:
+    """
+    NOTE: Users are not recommended to use any member functions of this class.
+    """
+
+    # A flag to indicate whether the evaluator is initialized complete.
+    _initialization_complete: bool
+    _hook: List[Hook]
+
+    def _init_optimizer_helpers(self, pure_model: Module | pl.LightningModule):
+        """
+        This is an internal API, ``pure_model`` means the model is the original model passed in by the user,
+        it should not be the modified model (wrapped, hooked, or patched by NNI).
+        That is, the optimizers & lr_schedulers obtained by ``Evaluator`` match the ``pure_model``.
+
+        This function is used to record the status of the optimizers & lr_schedulers,
+        and ensure NNI can reinitialize the optimizers & lr_schedulers with a similar but modified model.
+
+        NOTE: this is a part of Evaluator initialization, please make sure this function has been called before using other evaluator functions.
+        """
+        raise NotImplementedError
+
+    def bind_model(self, model: Module | pl.LightningModule, param_names_map: Dict[str, str] | None = None):
+        """
+        NOTE: NNI users don't need to call this function. This function usually called by NNI Compressor.
```

Comment:
I would suggest to remove all the NOTE message from member functions' docstring, only keep one in the class docstring